### PR TITLE
Change_map service to map_server [Rebase/Noetic]

### PR DESCRIPTION
**Pull request by DLu**
This is a redo of #461 for noetic, integrating changes that have been made in the last 4 years since the original PR. :roll_eyes: 

The CI will fail until the changes from https://github.com/ros/common_msgs/pull/152 are released, see https://github.com/ros/common_msgs/issues/167

The return codes from the service are not ideal (in that it only returns success or failure, not the more nuanced failure codes described in the service) but I was trying to keep close to @sloretz's original PR structure.

### CODE REVIEW

File: map_server/CMakeLists.txt [hunk 1]
```diff
@@ -4,6 +4,7 @@ project(map_server)
 find_package(catkin REQUIRED
         COMPONENTS
             roscpp
+            roslib
```
Comment (mikeferguson) [inline map_server/CMakeLists.txt]: this seems to only be in the Cmake/package.xml - where is it actually used/required?

File: map_server/src/main.cpp [hunk 2]
```diff
@@ -67,156 +68,218 @@ class MapServer
...
-            exit(-1);
-          }
+  private:
+    ros::NodeHandle n;
+    ros::Publisher map_pub;
+    ros::Publisher metadata_pub;
+    ros::ServiceServer service;
+    ros::ServiceServer change_map_srv_;
+    bool deprecated;
+    std::string frame_id_;
```
Comment (mikeferguson) [inline map_server/src/main.cpp:107]: should really be more consistent in naming here - member variable should all have the trailing underscore_

**Review by mikeferguson — CHANGES_REQUESTED** [review]

**Comment by EricWiener** [comment]
Any progress updates on this?

**Comment by DLu** [comment]
@ros-pull-request-builder retest this please

**Comment by EricWiener** [comment]
@DLu thanks for seeing if the tests pass. Does it look good to merge?

**Comment by DLu** [comment]
I think so. Let's get the +1 from @mikeferguson first since he had formally reviewed it earlier.

**Comment by EricWiener** [comment]
Any progress on this?

File: map_server/src/main.cpp [hunk 3]
```diff
@@ -67,156 +68,218 @@ class MapServer
...
+        response.result = response.RESULT_SUCCESS;
       }
-      catch (std::runtime_error e)
+      else
       {
-          ROS_ERROR("%s", e.what());
-          exit(-1);
+        response.result = response.RESULT_UNDEFINED_FAILURE;
       }
+      ROS_INFO("Changed map to %s", request.map_url.c_str());
```
Comment (mikeferguson) [inline map_server/src/main.cpp]: Seems odd to tell the user we changed the map if the loadMapFromYaml failed? Should this message be moved into the if() statement only? (it looks like we already get error messages from the loadMapFromYaml function).

**Review by mikeferguson — APPROVED** [review]

**Comment by engineeve** [comment]
Any chance this will be merged within the next month?

**Comment by HappySamuel** [comment]
Hi @DLu 

I saw that there's an addition feature (change_map) to the map_server. Is this add-on feature available if install via apt? 
`$ sudo apt install ros-noetic-map-server`
Because i check wiki of [ros noetic map server](http://wiki.ros.org/map_server), it remains outdated.

Best,
Samuel

**Comment by DLu** [comment]
No, and its been over a year since the last release, so its overdue.

**Pull request merged**